perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/MembersManagerBlImpl.java

```diff
@@ -1458,170 +1460,170 @@ public boolean canExtendMembershipWithReason(PerunSession sess, Member member) t
+		Pattern p = Pattern.compile("([0-9]+).([0-9]+).");
+		Matcher m = p.matcher(period);
+		if (m.matches()) {
+		  int day = Integer.valueOf(m.group(1));
+		  int month = Integer.valueOf(m.group(2));
+
+		  // Get current year
+		  int year = calendar.get(Calendar.YEAR);
+
+		  // We must detect if the extension date is in current year or in a next year
+		  boolean extensionInNextYear;
+		  Calendar extensionCalendar = Calendar.getInstance();
+		  extensionCalendar.set(year, month-1, day);
+		  Calendar today = Calendar.getInstance();
+		  if (extensionCalendar.before(today)) {
+			// Extension date is in a next year
+			extensionInNextYear = true;
+		  } else {
+			// Extension is in the current year
+			extensionInNextYear = false;
+		  }
+
+		  // Set the date to which the membershi should be extended, can be changed if there was grace period, see next part of the code
+		  calendar.set(year, month-1, day); // month is 0-based
+		  if (extensionInNextYear) {
+			calendar.add(Calendar.YEAR, 1);
+		  }
+
+		  // ***** GRACE PERIOD *****
+		  // Is there a grace period?
+		  if (membershipExpirationRules.get(MembersManager.membershipGracePeriodKeyName) != null) {
+			String gracePeriod = membershipExpirationRules.get(MembersManager.membershipGracePeriodKeyName);
+			// If the extension is requested in period-gracePeriod then extend to next period
+
+			// Get the value of the grace period
+			p = Pattern.compile("([0-9]+)([dmy]?)");
+			m = p.matcher(gracePeriod);
+			if (m.matches()) {
+			  String countString = m.group(1);
+			  int amount = Integer.valueOf(countString);
+
+			  // Set the gracePeriodCalendar to the extension date
+			  Calendar gracePeriodCalendar = Calendar.getInstance();
+			  gracePeriodCalendar.set(year, month-1, day);
+			  if (extensionInNextYear) {
+				gracePeriodCalendar.add(Calendar.YEAR, 1);
+			  }
+
+			  int field;
+			  String dmyString = m.group(2);
+			  if (dmyString.equals("d")) {
+				field = Calendar.DAY_OF_YEAR;
+			  } else if (dmyString.equals("m")) {
+				field = Calendar.MONTH;
+			  } else if (dmyString.equals("y")) {
+				field = Calendar.YEAR;
+			  } else {
+				throw new InternalErrorException("Wrong format of gracePeriod in VO membershipExpirationRules attribute. gracePeriod: " + gracePeriod);
+			  }
+			  // subtracts period definition, e.g. 3m
+			  gracePeriodCalendar.add(field, -amount);
+
+			  // Check if we are in grace period
+			  if (gracePeriodCalendar.before(Calendar.getInstance())) {
+				// We are in grace period, so extend to the next period
+				calendar.add(Calendar.YEAR, 1);
+			  }
+			}
+		  }
+		} else {
+		  throw new InternalErrorException("Wrong format of period in VO membershipExpirationRules attribute. Period: " + period);
+		}
+	  }
+
+	  // Reset hours, minutes and seconds to 0
+	  calendar.set(Calendar.HOUR, 0);
+	  calendar.set(Calendar.MINUTE, 0);
+	  calendar.set(Calendar.SECOND, 0);
+	  calendar.set(Calendar.MILLISECOND, 0);
+	}
+
+	return calendar.getTime();
```

Comment:
This fixing of code indentation is not need for this pull request. Please remove it.